flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,224 @@
+import org.apache.flink.runtime.io.network.buffer.LocalBufferPool;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FatalExitExceptionHandler;
+
+import org.apache.flink.shaded.guava30.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * The implementation for {@link TieredStorageMemoryManager}. This is to request or recycle buffers
+ * from {@link LocalBufferPool} for different memory owners, for example, the tiers, the buffer
+ * accumulator, etc.
+ *
+ * <p>Note that the memory owner should register its {@link TieredStorageMemorySpec} firstly before
+ * requesting buffers.
+ */
+public class TieredStorageMemoryManagerImpl implements TieredStorageMemoryManager {
+
+    /** Initial delay before checking flush. */
+    public static final int DEFAULT_CHECK_FLUSH_INITIAL_DELAY_MS = 10;
+
+    /** Check flush period. */
+    public static final int DEFAULT_CHECK_FLUSH_PERIOD_DURATION_MS = 50;
+
+    /** The tiered storage memory specs of each memory user owner. */
+    private final Map<Object, TieredStorageMemorySpec> tieredMemorySpecs;
+
+    /** The registered callbacks to flush the buffers in the registered tiered storages. */
+    private final List<Runnable> bufferFlushCallbacks;
+
+    /** The buffer pool usage ratio of triggering the registered storages to flush buffers. */
+    private final float numBuffersTriggerFlushRatio;
+
+    /**
+     * Indicate whether to start the buffer flush checker thread. If the memory manager is used in
+     * downstream, the field will be false because no buffer flush checker is needed.
+     */
+    private final boolean shouldStartBufferFlushChecker;
+
+    /** The number of requested buffers from {@link BufferPool}. */
+    private final AtomicInteger numRequestedBuffers;
+
+    /** A thread to check whether to flush buffers in each tiered storage. */
+    private ScheduledExecutorService executor;
+
+    /** The total number of guaranteed buffers for all tiered storages. */
+    private int numTotalGuaranteedBuffers;
+
+    /** The buffer pool where the buffer is requested or recyceld. */
+    private BufferPool bufferPool;
+
+    /**
+     * Indicate whether the {@link TieredStorageMemoryManagerImpl} is in running state. Before
+     * setting up, this field is false.
+     *
+     * <p>Note that before requesting buffers or getting the maximum allowed buffers, this running
+     * state should be checked.
+     */
+    private boolean isRunning;
```

Comment:
For `isRunning`, we should set it to `false` in `release()`. Or we can rename it to `initialized`.

Reply:
Fixed.